R/lik.R

```diff
@@ -55,46 +55,92 @@ logF_lik = function(df1,df2){
+         const = TRUE,
+         lcdfFUN = function(x){pgamma(abs(x),shape=y+1,rate=scale,log.p=TRUE)+y*log(scale)},
+         lpdfFUN = function(x){dgamma(abs(x),shape=y+1,rate=scale,log=TRUE)+y*log(scale)},
+         etruncFUN = function(a,b){-my_etruncgamma(-b,-a,y+1,scale)},
+         e2truncFUN = function(a,b){my_e2truncgamma(-b,-a,y+1,scale)},
+         data=list(y=y,link=link))
+  }else if (link=="log"){
+    y1 = y+1e-5 # add pseudocount
+    list(name = "pois",
+         const = TRUE,
+         lcdfFUN = function(x){pgamma(exp(-x),shape=y1,rate=scale,log.p=TRUE)-log(y1)+y*log(scale)},
+         lpdfFUN = function(x){dgamma(exp(-x),shape=y1,rate=scale,log=TRUE)-log(y1)+y*log(scale)},
+         etruncFUN = function(a,b){-my_etruncgamma(exp(-b),exp(-a),y1,scale)},
+         e2truncFUN = function(a,b){my_e2truncgamma(exp(-b),exp(-a),y1,scale)},
+         data=list(y=y,link=link))
+  }
 }
 
 #' @title Likelihood object for Binomial error distribution
 #' @description Creates a likelihood object for ash for use with Binomial error distribution
+#' @details Suppose we have Binomial observations \code{y} where \eqn{y_i\sim Bin(n_i,\p_i)}. 
+#'    We either put an unimodal prior g on the success probabilities \eqn{p_i\sim g} (by specifying 
+#'    \code{link="identity"}) or on the logit success probabilities \eqn{logit(p_i)\sim g} 
+#'    (by specifying \code{link="logit"}). Either way, ASH with this Binomial likelihood function 
+#'    will compute the posterior mean of the success probabilities \eqn{p_i}.
 #' @param y Binomial observations
 #' @param n Binomial number of trials
+#' @param link Link function. The "identity" link directly puts unimodal prior on Binomial success
+#'  probabilities p, and "logit" link puts unimodal prior on logit(p).
 #' 
 #' @examples 
 #'    p = rbeta(100,2,2) # prior mode: 0.5
 #'    n = rpois(100,10)
 #'    y = rbinom(100,n,p) # simulate Binomial observations
 #'    ash(rep(0,length(y)),1,lik=binom_lik(y,n))
 #' @export
-binom_lik = function(y,n){
-  list(name = "binom",
-       const = TRUE,
-       lcdfFUN = function(x){pbeta(abs(x),shape1=y+1,shape2=n-y+1,log.p=TRUE)-log(n+1)},
-       lpdfFUN = function(x){dbeta(abs(x),shape1=y+1,shape2=n-y+1,log=TRUE)-log(n+1)},
-       etruncFUN = function(a,b){-my_etruncbeta(-b,-a,y+1,n-y+1)},
-       e2truncFUN = function(a,b){my_e2truncbeta(-b,-a,y+1,n-y+1)},
-       data=list(y=y,n=n))
+binom_lik = function(y,n,link=c("identity","logit")){
+  link = match.arg(link)
+  if (link=="identity"){
+    list(name = "binom",
+         const = TRUE,
+         lcdfFUN = function(x){pbeta(abs(x),shape1=y+1,shape2=n-y+1,log.p=TRUE)-log(n+1)},
+         lpdfFUN = function(x){dbeta(abs(x),shape1=y+1,shape2=n-y+1,log=TRUE)-log(n+1)},
+         etruncFUN = function(a,b){-my_etruncbeta(-b,-a,y+1,n-y+1)},
+         e2truncFUN = function(a,b){my_e2truncbeta(-b,-a,y+1,n-y+1)},
+         data=list(y=y,n=n,link=link))
+  }else if(link=="logit"){
+    y1 = y+1e-5 # add pseudocount
```

Comment:
would it be possible to code the link function so that essentially any function f could be passed as the link? That is directly specify the  link function as a function, rather than naming it and then passing the name, and then having lots of if... then.. constructions?